### PR TITLE
Allow Dust Transaction Building + Mempool Rule To Reject

### PR DESCRIPTION
**Pull request by fassadlr**
As per agreement we will now allow the building of transactions with dust-like amounts but will reject via the mempool.

### CODE REVIEW

**Comment by maciejzaleski** [comment]
Here are the rules which we need to implement:
- We should allow any transaction which has `vin.Count > vout.Count`. This is a good thing as we are consolidating UTXOs and improving the network. The extreme case of this would be someone sending two UTXOs of value 0.00000001 to single address (ignore the fact that he will pay more in fees to what he will get). This is a good thing as we have reduced the set of UTXOs by one.
- If the transaction is of type OP_RETURN we should ignore transaction value check; we need this to avoid burning STRATs for cross chain transfers. Am I right thinking that we do not index OP_RETURN transactions anywhere?
- The logic needs to be different for Cirrus network where we are currently refunding gas which could be less than dust amount.
- We need to have a check on per output basis and not on the total value of transaction. If any output is less than dust we should reject the whole transaction.
- We should decide which output is a dust based on a common formula which we already have in the code. The fee needs to be less than 1/3 of the total payment amount.

**Comment by MithrilMan** [comment]
What's the reason you want to allow to create transactions with dust but not accept them in memorypool?

I don't see a valid use case for this.
- If we don't accept our built transaction in mempool it isn't going to be mined
- If we accept it, only ourself will be allowed to mine such transactions because other peers will reject it (not going in their mempool)

It seems to me we are wasting effort for something that isn't useufl.

> * If the transaction is of type OP_RETURN we should ignore transaction value check; we need this to avoid burning STRATs for cross chain transfers. Am I right thinking that we do not index OP_RETURN transactions anywhere?

Currently even with DustPrevention enabled, these transaction were passing because the check doesn't happens on OP_RETURN transactions

> * The logic needs to be different for Cirrus network where we are currently refunding gas which could be less than dust amount.

Can you elaborate on this?

> * We should decide which output is a dust based on a common formula which we already have in the code. The fee needs to be less than 1/3 of the total payment amount.

This is too simplicistic, the concept of Fee isn't straightforward and can't be computed by a simple formula, because what today isn't a dust, tomorrow may be and this should be taken into account.
Bitcoin has fee estimation based on time, something we may want to implement (even because our Fee estimation actually is fixed)

An ELI5 on Dust: an unspent coin (utxo) whose value is less of the cost needed to include it in a transaction.
Basically it's a dust when the utxo value <= fee cost to cover the size of the space used to include that utxo as input in the transaction.

The above doesn't take into consideration the fee increase over time (what I was talking about earlier), e.g. in bitcoin, where blocks are full, fee increase over time because of the fee market and the competition to have a transaction included in a specified timeframe (this is why if you want to see your transaction included in the very next blocks you'll set an higher fee respect to a transaction that's ok to be included in few hours).
This mean that the value to check against utxo value isn't current feeRate but it's an estimate of what a fee will be.

> * We should allow any transaction which has vin.Count > vout.Count. This is a good thing as we are consolidating UTXOs and improving the network

For the reason explained above, I don't agree 100% on this.
I agree that a transaction that shrink the UTXO set is positive, but if it generates outputs that are dust, these utxos are going to stay there "forever", so you are not really helping the nodes in the mid/long term

File: src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxTotalOutVsFeeRule.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Features.MemoryPool.Interfaces;
+
+namespace Stratis.Bitcoin.Features.MemoryPool.Rules
+{
+    /// <summary>
+    /// This rule ensures that the transaction's total out is at least equal to the network's minimum relay fee.
+    /// </summary>
+    public sealed class CheckTxTotalOutVsFeeRule : MempoolRule
```
Comment (MithrilMan) [inline src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxTotalOutVsFeeRule.cs]: We have already a rule to check fees: CheckFeeMempoolRule and the fee has to be checked against the size of the transaction, not the total output (what's the point of that check?)

Actually you are checking the sum of the output, compared to the min fee computed using the size of the transaction.
If you aim to check if an output is a dust, you should compare the value to the minrelayfee computed against the sum of the size of the inputs required to redeem such generated outputs

Anyway I don't think we need this rule as described in my comment

**Comment by dangershony** [comment]
>As per agreement we will now allow the building of transactions with dust-like amounts but will reject via the mempool.  

I thought this functionality as already implemented, when the mempool was ported the dust relay logic should have been ported as well.

**Comment by fassadlr** [comment]
@dangershony Thanks for referencing the PR on Bitcoin. Seeing as that is protocol change we will look at implementing the dust definition change on a separate PR.

After discussions with @maciejzaleski the implementation is as follows:

1. We do check for dust amounts in NBitcoin's `TransactionBuilder`. So if the user tries to submit a transaction with `DustPrevention` disabled, via the WalletManager it will get an error:

`
{Dust output detected, output value is 0.00001000, policy minimum is 0.00005460}
`

2. However, we feel that it is still necessary to check for this in the mempool as someone can submit a transaction to the mempool without calling the builder's `Verify` method.
3. We will ignore (pass) transactions which contain an `OpReturn` output.
4. We will not include this rule on the Cirrus Networks until we better understand how to handle dust amounts on the sidechain network.

File: src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+namespace Stratis.Bitcoin.Features.MemoryPool.Rules
+{
+    /// <summary>
+    /// This rule ensures that the transaction does not contain any outputs which are considered dust.
+    /// </summary>
+    public sealed class CheckTxOutDustRule : MempoolRule
+    {
+        public CheckTxOutDustRule(
+            Network network,
+            ITxMempool mempool,
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: ITxMempool is not needed anymore (not  a big deal though).

Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: Its needed by the other mempool rules ;)

File: src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        {
+        }
+
+        /// <inheritdoc />
+        public override void CheckTransaction(MempoolValidationContext context)
+        {
+            foreach (var txOut in context.Transaction.Outputs)
+            {
+                if (StandardTransactionPolicy.IsOpReturn(txOut.ScriptPubKey.ToBytes()))
+                    break;
```
Comment (maciejzaleski) [inline src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: That should be continue. If we break than someone could exploit this and send 1000 dust payments with a single payment to nulldata address.

File: src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            MempoolSettings mempoolSettings,
+            ChainIndexer chainIndexer,
+            ILoggerFactory loggerFactory) : base(network, mempool, mempoolSettings, chainIndexer, loggerFactory)
+        {
+        }
+
+        /// <inheritdoc />
+        public override void CheckTransaction(MempoolValidationContext context)
+        {
+            foreach (var txOut in context.Transaction.Outputs)
```
Comment (maciejzaleski) [inline src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: Before the loop we should check if `vin.Count > vout.Count`; reducing total set of UTXO is a good thing for the network regardless of the amounts involved.
`if (vin.Count > vout.Count) break;`

File: src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs (hunk 3 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: Ahh well spotted 👍

File: src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs (hunk 4 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: I don't see how this matter for dust.

File: src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs (hunk 3 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: Ah yes of course 👍

File: src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs (hunk 4 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: @dangershony the decision here is to reduce the amount of UTXOs if the inputs are more than the outputs. We are effectively consolidating dust utxo's on the network by allowing this transaction to pass.

Comment (dangershony) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: If you have 20 dust inputs and a dust output its still dust no?

Comment (dangershony) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: From the link above
```
Bitcoin Core considers a transaction output to be dust, when its value is lower than the cost of spending it at the dustRelayFee rate.
```

I think this should be the only indicator to disallow relay of dust and not number of outputs or inputs.

Comment (maciejzaleski) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: Yes, you are reducing 20 dust UTXO to one dust UTXO. As a result there is less UTXOs which is a good thing

Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: That is true but then we decided that it is user's issue then. If he/she wants to submit a transaction like that we should allow it, even it costs them more to send it. At least we are gaining from a smaller UTXO set.

This is a highly unlikely scenario though ;)

File: src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs [hunk 5]
```diff
@@ -22,10 +22,15 @@ public sealed class CheckTxOutDustRule : MempoolRule
         /// <inheritdoc />
         public override void CheckTransaction(MempoolValidationContext context)
         {
+            // If there are more inputs than outputs we don't care if they are dust transactions
+            // as we are effectively consolidating utxo's on the network.
+            if (context.Transaction.Inputs.Count > context.Transaction.Outputs.Count)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: I am not sure this is a good change.

Comment (dangershony) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: So to attack the network I create many small outputs that are just above dust, and then I cut that number by half to get two inputs and one dust output and the relay will allow this to propagate.

Comment (maciejzaleski) [reply src/Stratis.Bitcoin.Features.MemoryPool/Rules/CheckTxOutDustRule.cs]: I've done some analysis and I agree with @dangershony. Let's remove that change which I've proposed earlier.

**Pull request merged**